### PR TITLE
Palettes refactor (and some other stuff)

### PR DESCRIPTION
**Pull request by kanye-quest**
- Create a new `PaletteState` class to encapsulate global palette state
- Modify `VisualizationScene` and gl3 classes to use a local instance of `PaletteState`
- Move handle types into `gl/types.hpp` header
- Move common geometry functions into a single header

### CODE REVIEW

File: lib/sdl.cpp [hunk 1]
```diff
@@ -652,6 +652,7 @@ void SdlWindow::mainIter()
       {
          if (glvis_command->Execute() < 0)
          {
+            cout << "GlvisCommand signalled exit" << endl;
```
Comment (tzanio) [inline lib/sdl.cpp:655]: Styling should be `GLVisCommand` or `glvis_command`

File: lib/sdl.cpp [hunk 2]
```diff
@@ -663,7 +664,11 @@ void SdlWindow::mainIter()
       if (glvis_command && visualize == 1 &&
           (status = glvis_command->Execute()) != 1)
       {
-         if (status < 0) { running = false; }
+         if (status < 0)
+         {
+            cout << "GlvisCommand signalled exit" << endl;
```
Comment (tzanio) [inline lib/sdl.cpp:669]: See above styling comment

File: glvis.cpp [hunk 3]
```diff
@@ -160,7 +160,7 @@ bool GLVisInitVis(int field_type)
             vs->Zoom(1.8);
             // Use the 'bone' palette when visualizing a 2D mesh only (otherwise
             // the 'jet-like' palette is used in 2D, see vssolution.cpp).
-            paletteSet(4);
+            vs->GetPalette().SetPalette(4);
```
Comment (tzanio) [inline glvis.cpp:163]: This chain reads weird to me. Can it be just `vs->Palette()->Set(4)` or maybe an alternative shortcut `vs->SetPalette(4)`?

File: glvis.cpp [hunk 4]
```diff
@@ -703,14 +703,16 @@ void ExecuteScriptCommand()
          int pal;
          scr >> pal;
          cout << "Script: palette: " << pal << endl;
-         paletteSet(pal-1);
+         vs->GetPalette().SetPalette(pal);
```
Comment (tzanio) [inline glvis.cpp:706]: `pal` or `pal-1` here?

File: lib/geom_utils.hpp [hunk 5]
```diff
@@ -0,0 +1,147 @@
+// Copyright (c) 2010-2020, Lawrence Livermore National Security, LLC. Produced
```
Comment (tzanio) [inline lib/geom_utils.hpp:1]:
```suggestion
// Copyright (c) 2010-2021, Lawrence Livermore National Security, LLC. Produced
```

File: lib/geom_utils.hpp [hunk 6]
```diff
@@ -0,0 +1,147 @@
...
+// CONTRIBUTING.md for details.
+
+#ifndef GLVIS_GEOM_UTILS_HPP
+#define GLVIS_GEOM_UTILS_HPP
+#include "mfem.hpp"
+
+// Some inline functions
+
+inline void LinearCombination(const double a, const double x[],
+                              const double b, const double y[], double z[])
```
Comment (tzanio) [inline lib/geom_utils.hpp:19]: Were all the functions in this file just moved (without changes) from other files?

**Review by tzanio — COMMENTED** [review]
I didn't track all code moved between files, but in general this looks correct and appears to work correctly :)

Please document in `CHANGELOG`

File: lib/geom_utils.hpp (hunk 6 above)
Comment (kanye-quest) [reply lib/geom_utils.hpp:19]: Yes, these are just a straight copy-and-paste.

File: lib/sdl.cpp (hunk 1 above)
Comment (kanye-quest) [reply lib/sdl.cpp]: Not actually sure how this addition made it in, so I just removed it.

File: lib/sdl.cpp (hunk 2 above)
Comment (kanye-quest) [reply lib/sdl.cpp]: Same as above

File: glvis.cpp (hunk 4 above)
Comment (kanye-quest) [reply glvis.cpp]: I think this should be the original `pal-1` value. Fixed.

File: glvis.cpp (hunk 3 above)
Comment (kanye-quest) [reply glvis.cpp]: I tried:
- removing `GetPalette()` for a `VisualizationScene::palette` global variable
- renaming `PaletteState::{Get,Set,Next,Prev}Palette` to `PaletteState::{Get,Set,Next,Prev}Index`

Let me know if that works for you.

Comment (tzanio) [reply glvis.cpp]: Looks good, thanks!

**Review by tzanio — APPROVED** [review]

**Comment by tzanio** [comment]
@tomstitt can you take a look when you get a chance?

**Review by tomstitt — COMMENTED** [review]
I didn't look over this carefully since I don't know this part of code well but I fixed and tested the Emscripten build and that looks good now so I'm happy with it!

**Review by tomstitt — APPROVED** [review]

**Comment by tzanio** [comment]
This PR resulted in a number of warnings in MFEM's nightly regression tests.

See for example the log at https://mybitbucket.llnl.gov/projects/MFEM/repos/autotest/browse/tux375/2021-02-27-llnl-gh-next/build-test-basic.err, with warnings like:

```
lib/gl/../palettes.hpp:18:7: warning: ‘PaletteState’ has a field ‘PaletteState::palette_tex’ whose type uses the anonymous namespace
lib/gl/../palettes.hpp:18:7: warning: ‘PaletteState’ has a field ‘PaletteState::alpha_tex’ whose type uses the anonymous namespace
lib/gl/renderer.hpp:97:7: warning: ‘gl3::GLDevice’ has a field ‘gl3::GLDevice::passthrough_texture’ whose type uses the anonymous namespace
lib/gl/renderer_core.hpp:58:11: warning: ‘gl3::CoreGLDevice::VBOData’ has a field ‘gl3::CoreGLDevice::VBOData::vert_buf’ whose type uses the anonymous namespace
lib/gl/renderer_core.hpp:19:7: warning: ‘gl3::CoreGLDevice’ has a field ‘gl3::CoreGLDevice::feedback_prgm’ whose type uses the anonymous namespace
lib/gl/renderer_core.hpp:19:7: warning: ‘gl3::CoreGLDevice’ has a field ‘gl3::CoreGLDevice::global_vao’ whose type uses the anonymous namespace
lib/gl/renderer_core.hpp:19:7: warning: ‘gl3::CoreGLDevice’ has a field ‘gl3::CoreGLDevice::feedback_vbo’ whose type uses the anonymous namespace
lib/gl/renderer_core.hpp:19:7: warning: ‘gl3::CoreGLDevice’ has a field ‘gl3::CoreGLDevice::default_prgm’ whose type uses the anonymous namespace
```

We should really add CI and proper testing for GLVis!

**Pull request merged**